### ISSUE TITLE
setup: add command line option to specify the answer to each question

### ISSUE THREAD

**Issue by shtrom** [asamuzaK/withExEditorHost#36]
Something like the following would be useful for automated (scripted) installs:

    node index.js --setup --browser=firefox \ # already supported
        --config-path=... \ # maybe already supported
        --overwrite-existing \ # not apparently supported
        --editor-path=... \ # ditto
        --editor-options=... \ # ditto
        --editor-add-file-path=... # ditto

**Comment by asamuzaK** [comment]
Makes sense.
Thanks for proposal.

As for overwrite, I am asking the user twice.
The first time is whether to overwrite the config setting if it exists or not, and second is whether to overwrite the editor config.
Do you think users can understand what command line is meant, for example, `--overwrite-config` and `--overwrite-editor-config`?

**Comment by asamuzaK** [comment]
One more.
I feel somewhat ambiguous as to what we are going to do with `--editor-add-file-path`.
I am not a native English speaker, can you propose a better command arg which tells user whether to `put the file's path at the end of the command`?

**Comment by shtrom** [comment]
On Wed 04 Jul 2018 at 09:13:13 -0700, asamuzaK (Kazz) wrote:
> Do you think users can understand what command line is meant, for example, `--overwrite-config` and `--overwrite-editor-config`?

Yep, this sounds good to me.

We may also want a `--no-overwrite`, actually, so it does nothing if any
config is already there.

On Wed 04 Jul 2018 at 09:18:33 -0700, asamuzaK (Kazz) wrote:
> I feel somewhat ambiguous as to what we are going to do with `--editor-add-file-path`.
> I am not a native English speaker, can you propose a better command arg which tells user whether to `put the file's path at the end of the command`?

Actually, I'm not sure what this option does, even when asked
interactively. Could you clarify this for me?

Does it
* add the filename at the end of the editor command, or
* add the _full path_ to the relative filename that is unconditionally added?

I'm not sure in which case this is useful. My thoughts would be to just
unconditionally add the full path to the file, but perhaps I'm missing
something?

-- 
Olivier Mehani <shtrom@ssji.net>
PGP fingerprint: 4435 CF6A 7C8D DD9B E2DE  F5F9 F012 A6E2 98C6 6655
Confidentiality cannot be guaranteed on emails sent or received unencrypted.

**Comment by asamuzaK** [comment]
> Does it
> * add the filename at the end of the editor command, or
> * add the _full path_ to the relative filename that is unconditionally added? 

The former.
Some editor requires to put file path at the end of args like `foo bar <filepath>`

Still in progress, how about it?

```
> node index --help

  Usage: index [options] [command]

  Options:

    -v, --version                  output the version number
    -b, --browser <name>           specify the browser
    -c, --config-path <path>       path to save config files
    -o, --overwrite-config         overwrite config if exists
    -O, --overwrite-editor-config  overwrite editor config if exists
    -e, --editor-path <path>       editor path
    -a, --editor-args <args>       editor command args, need to be quoted
    -f, --file-after-args          put file path at the end of command args
    -h, --help                     output usage information

  Commands:

    setup|s                        run setup
```

**Comment by shtrom** [comment]
On Thu 05 Jul 2018 at 18:36:51 -0700, asamuzaK (Kazz) wrote:
> > Does it
> > * add the filename at the end of the editor command, or
> > * add the _full path_ to the relative filename that is unconditionally added? 
> The former.
> Some editor requires to put file path at the end of args like `foo bar <filepath>`

I still don't get it. What would the full command line look like?

I would expect something like this as the default

1. `/path/to/editor <args> filename`

or, better,

2. `/path/to/editor <args> /path/to/filename`

but does it do something like this?

3. `/path/to/editor <args> filename /path/to`

or is is just a way to avoid

4. `/path/to/editor /path/to/filename <args>`

In case it's trying to avoid 4., I'm wondering whether _any_ editor
would refuse 2.. If not, this could just be made the default, and the
option completely removed?

> Still in progress, how about it?
> 
> ```
> > node index --help
> 
>   Usage: index [options] [command]
> 
>   Options:
> 
>     -v, --version                  output the version number
>     -b, --browser <name>           specify the browser
>     -c, --config-path <path>       path to save config files
>     -o, --overwrite-config         overwrite config if exists
>     -O, --overwrite-editor-config  overwrite editor config if exists
>     -e, --editor-path <path>       editor path
>     -a, --editor-args <args>       editor command args, need to be quoted
>     -f, --file-after-args          put file path at the end of command args
>     -h, --help                     output usage information

Ah! Yeah, that looks good!

-- 
Olivier Mehani <shtrom@ssji.net>
PGP fingerprint: 4435 CF6A 7C8D DD9B E2DE  F5F9 F012 A6E2 98C6 6655
Confidentiality cannot be guaranteed on emails sent or received unencrypted.

**Comment by asamuzaK** [comment]
default:
`path/to/editor path/to/file args`

if `-f` is given or `y` in command prompt:
`path/to/editor args path/to/file`

**Comment by shtrom** [comment]
On Fri 06 Jul 2018 at 19:00:30 -0700, asamuzaK (Kazz) wrote:
> default:
> `path/to/editor path/to/file args`
> 
> if `-f` is given or `y` in command prompt:
> `path/to/editor args path/to/file`

Cool, so that's my question. If we make the `-f` behaviour the default,
would any editor fail because they require the current default? AFAIK
most editors would be able to handle it.

-- 
Olivier Mehani <shtrom@ssji.net>
PGP fingerprint: 4435 CF6A 7C8D DD9B E2DE  F5F9 F012 A6E2 98C6 6655
Confidentiality cannot be guaranteed on emails sent or received unencrypted.

**Comment by asamuzaK** [comment]
> Cool, so that's my question. If we make the `-f` behaviour the default, would any editor fail because they require the current default? AFAIK most editors would be able to handle it.

OK, let's remove it.

**Comment by asamuzaK** [comment]
Filed #38

**Comment by asamuzaK** [comment]
I'll create a pre-release after #38 is fixed.
Please wait until then.

**Comment by asamuzaK** [comment]
Try build [Releases · asamuzaK/withExEditorHost](https://github.com/asamuzaK/withExEditorHost/releases "Releases · asamuzaK/withExEditorHost")

Note that command to run setup is `setup`, not `--setup` anymore.

```
> index setup -b firefox -o
```

**Comment by shtrom** [comment]
I'm finally getting down to testing it. It works well, except for one detail: I don't seem to be able to specify that I want no editor args from the command line:

```
npm run setup -- --overwrite-config --browser=firefox --editor-path $(which gvim) --overwrite-editor-config --editor-args "''"

> withexeditorhost@4.0.2 setup /home/shtrom/bordel/default-env/vendor/withExEditorHost
> npm install --production && node ./index.js setup "--overwrite-config" "--browser=firefox" "--editor-path" "/usr/bin/gvim" "--overwrite-editor-config" "--editor-args" "''"

audited 3960 packages in 3.911s
found 0 vulnerabilities

Created: /home/shtrom/.config/withexeditorhost/config/firefox
Created: /home/shtrom/.config/withexeditorhost/config/firefox/withexeditorhost.sh
Created: /home/shtrom/.mozilla/native-messaging-hosts/withexeditorhost.json
Enter command line options:
```
(Same with `"''"`)

When I do provide a dummy argument, it works well:
```
npm run setup -- --overwrite-config --browser=firefox --editor-path $(which gvim) --overwrite-editor-config --editor-args "a"

> withexeditorhost@4.0.2 setup /home/shtrom/bordel/default-env/vendor/withExEditorHost
> npm install --production && node ./index.js setup "--overwrite-config" "--browser=firefox" "--editor-path" "/usr/bin/gvim" "--overwrite-editor-config" "--editor-args" "a"

audited 3960 packages in 3.935s
found 0 vulnerabilities

Created: /home/shtrom/.config/withexeditorhost/config/firefox
Created: /home/shtrom/.config/withexeditorhost/config/firefox/withexeditorhost.sh
Created: /home/shtrom/.mozilla/native-messaging-hosts/withexeditorhost.json
Created: /home/shtrom/.config/withexeditorhost/config/firefox/editorconfig.json
```

It would also be good to have an updated help:
```
npm run setup --  --help                                                                                                     0s  master

> withexeditorhost@4.0.2 setup /home/shtrom/bordel/default-env/vendor/withExEditorHost
> npm install --production && node ./index.js setup "--help"

audited 3960 packages in 3.949s
found 0 vulnerabilities


  Usage: setup|s [options]

  run setup

  Options:

    -h, --help  output usage information
```

**Comment by asamuzaK** [comment]
Would you file a new issue for the editor args?
Thanks.
